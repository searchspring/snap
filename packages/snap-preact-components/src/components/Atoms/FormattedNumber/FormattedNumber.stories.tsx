import { h } from 'preact';

import { ArgsTable, PRIMARY_STORY, Markdown } from '@storybook/blocks';

import { FormattedNumber, FormattedNumberProps } from './FormattedNumber';
import { componentArgs, highlightedCode } from '../../../utilities';
import Readme from '../FormattedNumber/readme.md';

export default {
	title: 'Atoms/FormattedNumber',
	component: FormattedNumber,
	tags: ['autodocs'],
	parameters: {
		docs: {
			page: () => (
				<div>
					<Markdown
						options={{
							overrides: {
								code: highlightedCode,
							},
						}}
					>
						{Readme}
					</Markdown>
					<ArgsTable story={PRIMARY_STORY} />
				</div>
			),
		},
	},
	argTypes: {
		value: {
			description: 'Numeric value to be formatted',
			type: { required: true },
			table: {
				type: {
					summary: 'number',
				},
			},
			control: { type: 'number' },
		},
		symbol: {
			defaultValue: '',
			description: 'Unit symbol',
			table: {
				type: {
					summary: 'string',
				},
			},
			control: { type: 'text' },
		},
		symbolAfter: {
			defaultValue: true,
			description: 'Place unit symbol after the value',
			table: {
				type: {
					summary: 'boolean',
				},
				defaultValue: { summary: true },
			},
			control: { type: 'boolean' },
		},
		decimalPlaces: {
			defaultValue: 3,
			description: 'Number of decimal places',
			table: {
				type: {
					summary: 'number',
				},
				defaultValue: { summary: 3 },
			},
			control: { type: 'number' },
		},
		padDecimalPlaces: {
			defaultValue: true,
			description: 'Pad decimal places with zeros',
			table: {
				type: {
					summary: 'boolean',
				},
				defaultValue: { summary: true },
			},
			control: { type: 'boolean' },
		},
		thousandsSeparator: {
			defaultValue: '',
			description: 'Character used to separate thousands',
			table: {
				type: {
					summary: 'string',
				},
				defaultValue: { summary: '' },
			},
			control: { type: 'text' },
		},
		decimalSeparator: {
			defaultValue: '.',
			description: 'Character used to separate decimal values',
			table: {
				type: {
					summary: 'string',
				},
				defaultValue: { summary: '.' },
			},
			control: { type: 'text' },
		},
		raw: {
			description: 'Returns raw value without wrapping DOM node',
			defaultValue: false,
			table: {
				type: {
					summary: 'boolean',
				},
				defaultValue: { summary: false },
			},
			control: { type: 'boolean' },
		},
		...componentArgs,
	},
};

export const Default = (args: FormattedNumberProps) => <FormattedNumber {...args} />;
Default.args = {
	value: 1099.99,
};

export const Temperature = (args: FormattedNumberProps) => <FormattedNumber {...args} />;
Temperature.args = {
	value: 100,
	symbol: ' Â°C',
	decimalPlaces: 2,
};

export const Length = (args: FormattedNumberProps) => <FormattedNumber {...args} />;
Length.args = {
	value: 100,
	symbol: ' mm',
	decimalPlaces: 2,
};
