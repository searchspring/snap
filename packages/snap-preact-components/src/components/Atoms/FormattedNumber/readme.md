## Formatted Number

Renders a span that calls the formatNumber function in the Snap Toolbox. Has tons of props to allow you to format the number for all scenarios. price, strike price, Temperature, length, weight. 

## Additional Info

Uses the formatNumber function in the Snap Toolbox (https://github.com/searchspring/snap-toolbox/blob/main/filters/src/formatNumber.ts)

## Usage

 Default

``` jsx
<FormattedNumber value={1099.99} />
```
 Using custom symbol and setting decimal places

``` jsx
<FormattedNumber value={100} symbol={' Â°C'} decimalPlaces={2} />
```

