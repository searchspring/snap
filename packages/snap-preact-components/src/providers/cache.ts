export { CacheProvider } from '@emotion/react';
import createCache from '@emotion/cache';

export const cache = createCache({
	key: 'ss',
	prepend: true,
});
