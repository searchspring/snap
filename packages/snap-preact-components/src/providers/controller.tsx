import { jsx } from '@emotion/react';
import preact, { h, createContext, ComponentChildren, ComponentType } from 'preact';
import { useContext } from 'preact/hooks';
import type { AbstractController } from '@searchspring/snap-controller';

const ControllerContext = createContext<null | AbstractController>(null);

export const ControllerProvider = ({ children, controller }: { children: ComponentChildren; controller: AbstractController }) => {
	return <ControllerContext.Provider value={controller}>{children}</ControllerContext.Provider>;
};

export const useController = () => useContext(ControllerContext);

export const withController = (Component: ComponentType) => (props: any) => {
	return <Component controller={useController()} {...props} />;
};
