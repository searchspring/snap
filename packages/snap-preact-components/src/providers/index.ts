export * from './store';
export * from './theme';
