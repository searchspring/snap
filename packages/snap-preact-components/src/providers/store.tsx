import preact, { h, createContext, ComponentChildren, ComponentType } from 'preact';
import { useContext } from 'preact/hooks';
import type { AbstractStore } from '@searchspring/snap-store-mobx';

const StoreContext = createContext<null | AbstractStore>(null);

export const StoreProvider = ({ children, store }: { children: ComponentChildren; store: AbstractStore }) => {
	return <StoreContext.Provider value={store}>{children}</StoreContext.Provider>;
};

export const useStore = () => useContext(StoreContext);

export function withStore<C extends ComponentType>(Component: C): C {
	return ((props: any) => <StoreContext.Consumer>{(store) => <Component {...props} store={store} />}</StoreContext.Consumer>) as C;
}
