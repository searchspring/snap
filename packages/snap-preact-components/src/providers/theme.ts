import { ThemeProvider as EmotionTheme } from '@emotion/react';
export { css, useTheme, withTheme } from '@emotion/react';

interface ThemeProviderProps {
	theme: Partial<Theme> | ((outerTheme: Theme) => Theme);
	children?: any;
}
interface ThemeProvider {
	(props: ThemeProviderProps): React.ReactElement;
}
export const ThemeProvider: ThemeProvider = EmotionTheme;

export const defaultTheme: Theme = {
	colors: {
		primary: '#3A23AD',
		secondary: '#00cee1',
		hover: '#f8f6fd',
		text: {
			secondary: '#ffffff',
		},
		message: {
			error: '#cc1212',
			warning: '#ecaa15',
			info: '#4c3ce2',
		},
	},
	components: {},
};
export interface Theme {
	variables?: {
		[key: string]: unknown;
	};
	colors?: {
		primary: string;
		secondary: string;
		hover?: string;
		text?: {
			secondary: string;
		};
		message?: {
			error: string;
			warning: string;
			info: string;
		};
	};
	components?: any;
	namedComponents?: any;
}
