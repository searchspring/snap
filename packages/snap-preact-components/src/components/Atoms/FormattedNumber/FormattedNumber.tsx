/** @jsx jsx */
import { h } from 'preact';

import { jsx, css } from '@emotion/react';
import { filters } from '@searchspring/snap-toolbox';
import classnames from 'classnames';

import { Theme, useTheme } from '../../../providers/theme';
import { ComponentProps } from '../../../types';

const CSS = {
	formattedNumber: ({ style }) =>
		css({
			...style,
		}),
};

export function FormattedNumber(properties: FormattedNumberProps): JSX.Element {
	const globalTheme: Theme = useTheme();

	const props: FormattedNumberProps = {
		// default props
		symbol: '',
		decimalPlaces: 3,
		padDecimalPlaces: true,
		thousandsSeparator: '',
		decimalSeparator: '.',
		symbolAfter: true,
		// global theme
		...globalTheme?.components?.formattedNumber,
		// props
		...properties,
		...properties.theme?.components?.formattedNumber,
	};

	const {
		value,
		symbol,
		decimalPlaces,
		padDecimalPlaces,
		thousandsSeparator,
		decimalSeparator,
		symbolAfter,
		disableStyles,
		className,
		style,
	} = props;

	return (
		<span className={classnames('ss-formattednumber', className)} css={!disableStyles && CSS.formattedNumber({ style })}>
			{filters.formatNumber(value, {
				symbol,
				decimalPlaces,
				padDecimalPlaces,
				thousandsSeparator,
				decimalSeparator,
				symbolAfter,
			})}
		</span>
	);
}
export interface FormattedNumberProps extends ComponentProps {
	value: number;
	symbol?: string;
	decimalPlaces?: number;
	padDecimalPlaces?: boolean;
	thousandsSeparator?: string;
	decimalSeparator?: string;
	symbolAfter?: boolean;
}
