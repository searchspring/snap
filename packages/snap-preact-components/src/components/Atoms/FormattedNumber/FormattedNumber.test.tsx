import { h } from 'preact';
import { render } from '@testing-library/preact';

import { FormattedNumber } from './FormattedNumber';
import { ThemeProvider } from '../../../providers';

describe('FormattedNumber Component', () => {
	it('renders', () => {
		const rendered = render(<FormattedNumber value={1099.99} />);
		const formattednumberElement = rendered.container.querySelector('.ss__formatted-number');
		expect(formattednumberElement).toBeInTheDocument();
	});

	it('has default formatting', () => {
		const args = {
			value: 1099.99999,
			symbol: 'mm',
		};
		const rendered = render(<FormattedNumber {...args} />);
		const formattednumberElement = rendered.container.querySelector('.ss__formatted-number');
		const formattednumber = formattednumberElement?.textContent;
		expect(formattednumber).toBe('1099.999mm');
	});

	it('has support for custom options', () => {
		const args = {
			value: 1099.999999,
			symbol: ' £',
			decimalPlaces: 2,
			thousandsSeparator: '.',
			decimalSeparator: ',',
			symbolAfter: true,
		};
		const rendered = render(<FormattedNumber {...args} />);
		const formattednumberElement = rendered.container.querySelector('.ss__formatted-number');
		const formattednumber = formattednumberElement?.textContent;
		expect(formattednumber).toBe('1.099,99 £');
	});

	it('pads digits when it needs to', () => {
		const args = {
			value: 1099,
		};
		const rendered = render(<FormattedNumber {...args} />);
		const formattednumberElement = rendered.container.querySelector('.ss__formatted-number');
		const formattednumber = formattednumberElement?.textContent;
		expect(formattednumber).toBe('1099.000');
	});

	it('does not pads digits when prop specified', () => {
		const args = {
			value: 1099,
			padDecimalPlaces: false,
		};
		const rendered = render(<FormattedNumber {...args} />);
		const formattednumberElement = rendered.container.querySelector('.ss__formatted-number');
		const formattednumber = formattednumberElement?.textContent;
		expect(formattednumber).toBe('1099');
	});

	it('renders with additional style using prop', () => {
		const style = {
			padding: '20px',
		};

		const rendered = render(<FormattedNumber value={100} style={style} />);
		const formattednumberElement = rendered.container.querySelector('.ss__formatted-number')!;
		const styles = getComputedStyle(formattednumberElement);

		expect(styles.padding).toBe(style.padding);
	});

	it('can disable styling', () => {
		const args = {
			value: 1099.99,
			disableStyles: true,
			style: {
				padding: '20px',
			},
		};
		const rendered = render(<FormattedNumber {...args} />);
		const formattednumberElement = rendered.container.querySelector('.ss__formatted-number')!;
		const styles = getComputedStyle(formattednumberElement);
		expect(styles.background).not.toBe(args.style.padding);
	});

	it('uses custom className', () => {
		const args = {
			value: 1099.99,
			className: 'custom-class',
		};
		const rendered = render(<FormattedNumber {...args} />);
		const formattednumberElement = rendered.container.querySelector('.ss__formatted-number');
		expect(formattednumberElement?.classList).toContain(args.className);
	});

	it('is themeable with ThemeProvider', () => {
		const globalTheme = {
			components: {
				formattedNumber: {
					className: 'theme-class',
				},
			},
		};

		const rendered = render(
			<ThemeProvider theme={globalTheme}>
				<FormattedNumber value={33.333} />
			</ThemeProvider>
		);

		const formattednumberElement = rendered.container.querySelector('.ss__formatted-number');
		expect(formattednumberElement).toHaveClass(globalTheme.components.formattedNumber.className);
	});

	it('is themeable with theme prop', () => {
		const propTheme = {
			components: {
				formattedNumber: {
					className: 'theme-class',
				},
			},
		};

		const rendered = render(<FormattedNumber theme={propTheme} value={33.333} />);

		const formattednumberElement = rendered.container.querySelector('.ss__formatted-number');
		expect(formattednumberElement).toHaveClass(propTheme.components.formattedNumber.className);
	});

	it('is themeable and theme prop overrides ThemeProvider', () => {
		const globalTheme = {
			components: {
				formattedNumber: {
					className: 'global-class',
				},
			},
		};

		const propTheme = {
			components: {
				formattedNumber: {
					className: 'theme-class',
				},
			},
		};

		const rendered = render(
			<ThemeProvider theme={globalTheme}>
				<FormattedNumber value={33.333} theme={propTheme} />
			</ThemeProvider>
		);

		const formattednumberElement = rendered.container.querySelector('.ss__formatted-number');
		expect(formattednumberElement).toHaveClass(propTheme.components.formattedNumber.className);
		expect(formattednumberElement).not.toHaveClass(globalTheme.components.formattedNumber.className);
	});
});
