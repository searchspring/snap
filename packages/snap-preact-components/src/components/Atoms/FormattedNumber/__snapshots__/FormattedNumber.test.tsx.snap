// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormattedNumber Component can disable styling 1`] = `
<DocumentFragment>
  .emotion-0 {
  padding: 20px;
}

<span
    class="ss__formatted-number emotion-0"
  >
    1099.990
  </span>
</DocumentFragment>
`;

exports[`FormattedNumber Component does not pads digits when prop specified 1`] = `
<DocumentFragment>
  <span
    class="ss__formatted-number emotion-0"
  >
    1099
  </span>
</DocumentFragment>
`;

exports[`FormattedNumber Component has default formatting 1`] = `
<DocumentFragment>
  <span
    class="ss__formatted-number emotion-0"
  >
    1099.999mm
  </span>
</DocumentFragment>
`;

exports[`FormattedNumber Component has support for custom options 1`] = `
<DocumentFragment>
  <span
    class="ss__formatted-number emotion-0"
  >
    1.099,99 Â£
  </span>
</DocumentFragment>
`;

exports[`FormattedNumber Component is themeable and theme prop overrides ThemeProvider 1`] = `
<DocumentFragment>
  <span
    class="ss__formatted-number theme-class emotion-0"
  >
    33.333
  </span>
</DocumentFragment>
`;

exports[`FormattedNumber Component is themeable with ThemeProvider 1`] = `
<DocumentFragment>
  <span
    class="ss__formatted-number theme-class emotion-0"
  >
    33.333
  </span>
</DocumentFragment>
`;

exports[`FormattedNumber Component is themeable with theme prop 1`] = `
<DocumentFragment>
  <span
    class="ss__formatted-number theme-class emotion-0"
  >
    33.333
  </span>
</DocumentFragment>
`;

exports[`FormattedNumber Component pads digits when it needs to 1`] = `
<DocumentFragment>
  <span
    class="ss__formatted-number emotion-0"
  >
    1099.000
  </span>
</DocumentFragment>
`;

exports[`FormattedNumber Component renders 1`] = `
<DocumentFragment>
  <span
    class="ss__formatted-number emotion-0"
  >
    1099.990
  </span>
</DocumentFragment>
`;

exports[`FormattedNumber Component renders with additional style using prop 1`] = `
<DocumentFragment>
  .emotion-0 {
  padding: 20px;
}

<span
    class="ss__formatted-number emotion-0"
  >
    100.000
  </span>
</DocumentFragment>
`;

exports[`FormattedNumber Component uses bocachica theme 1`] = `
<DocumentFragment>
  <span
    class="ss__formatted-number emotion-0"
  >
    1099.990
  </span>
</DocumentFragment>
`;

exports[`FormattedNumber Component uses custom className 1`] = `
<DocumentFragment>
  <span
    class="ss__formatted-number custom-class emotion-0"
  >
    1099.990
  </span>
</DocumentFragment>
`;
